clash-lib/src/Clash/Core/TermLiteral.hs

```diff
@@ -62,11 +74,51 @@ import           Clash.Core.TermLiteral.TH
 showType :: TermLiteral a => Proxy a -> String
 showType proxy = showsTypePrec 0 proxy ""
 
+-- | Convert 'Term' to the constant it represents. Will return an error if
+-- (one of the subterms) fail to translate. It takes a 'BindingMap' as an
+-- argument, which is used to inline global binders. If you're using this
+-- function in a @BlackBoxFunction@/@NetlistMonad@ context, consider using
+-- 'termToDataM'.
+termToData
+  :: (HasCallStack, TermLiteral a)
+  => BindingMap
+  -> Term
+  -> Either Term a
+termToData bindingMap term = termToData# (inlineGlobalBinders bindingMap term)
+
+-- | Convert 'Term' to the constant it represents. Will return an error if
+-- (one of the subterms) fail to translate.
+termToDataM :: (HasCallStack, TermLiteral a) => Term -> NetlistMonad (Either Term a)
+termToDataM term = do
+  netlistState <- get
+  pure (termToData (netlistState ^. bindings) term)
+
+-- | Replace any global binders with their corresponding term
+inlineGlobalBinders :: BindingMap -> Term -> Term
+inlineGlobalBinders binders = go
+ where
+  go = \case
+    Var v
+      | isGlobalId v
+      , Just (bindingTerm -> t) <- lookupVarEnv v binders -> go t
```

Comment:
Perhaps `go` should take an integer to stop infinite loops here. (Though TBF `TermLiteral` isn't supposed to "parse" infinitely self recursive data types.)